### DISCUSSION TITLE
Logging

### DISCUSSION THREAD

**Discussion by brunetton** [jejellyroll-fr/fpdb-3#67] [development]
Hi @jejellyroll-fr 

I've seen that you've added quite a lot of new loggers (for example `handHistoryConverter`, `tourgraphViewer`, `tourplayerstats`, etc). It makes default console output very verbose, as there is no default configuration for those loggers in `logging.conf`.

Is this intentional ? Or do you like to have tons of debug outputs while coding ? Or do you use another way to select who can speak ? (in this case we should remove logging.conf, but it's used in `Configuration.py` so I presume it's used, I didn't looked deeper)

Thanks

**Comment by brunetton** [comment]
Okay, looking deeper I found that `logging.conf` is not used anymore. This explains why my logging configuration became totally broken.

If I understand well, you added a custom logger in `loggingFpdb.py`. I didn't looked deep but if I understand well why this has been added:
- you wanted to split log files based on time and size (this is why `TimedSizedRotatingFileHandler`)
- add colorization of strings
- do "something" in https://github.com/jejellyroll-fr/fpdb-3/commit/86deafa#diff-881d24502aa4b80dbc9a6e1d2348f882772d0c94d0ca63c3fc63e29107b158f1R102 but ... what ? I'm not sure but I've the feeling that you want to remove from stack all `loggingFpdb.py` code. Not sure is we could use existing code in logging ?

Sorry to bother but again but
- newcomers must follow the code and understand quite quickly what's going on and where (and logging is the very first tool that we use when we're new to a project)
- at any price we should **delete dead code** ! This is probably the most important thing to prevent a huge technical debt 
   - so if logging.conf is not used anymore, delete it, but IMHO we should adapt this logger to use this config file, and add new loggers to this file

Hope you don't hate me :D Thanks :)

**Reply by jejellyroll-fr** [reply]
hello Bruno,

yes, you're right ... once again I went too fast ... I'm starting to get a lot of private requests for improvements:

for the heritage of the logging class, you're right ... I just didn't take the time to read fully the lib doc ... to do something clean

yes, you're right about the dead code ... I'm going too fast ;)

I confess I didn't originally intend to improve the desktop application, it's not my IT culture, I wanted to make it a web app ... but I've got requests for evolution


there are several points I'd like to address:
redo at least the class diagrams, probably with Mermaid, because there are a lot of bugs when you look at the details - for example, I've just spent several days on the Hand and Pot classes, and they deserve clearer documentation
even if I'd started by writing the tests, I'm not sure I'd have generated regression elsewhere)
for me if you look, in hand.py, there is a problem of conception,  there should be a compositional relationship in Hand and Pot: with clearer role and responsabilty between the 2 classes
should something like this:
```mermaid
classDiagram
    class Hand {
        +config : dict <<from Config>>
        +sitename : str
        +gametype : dict
        +handText : str
        +builtFrom : None | str
        +pot : Pot <<sub-element>>
        +players : List[Player]
        +__init__(config: dict, sitename: str, gametype: dict, handText: str, builtFrom: None|str)
        +addPlayer(seat, name, chips, position, sitout, bounty)
        +removePlayer(name)
        +updateTourneyResults(db)
        +printHand()
    }

    class Pot {
        +players : List[Player] <<delegated>>
        +totalCollected : float
        +rake : float
        +__init__()
        +addMoney(player: Player, amount: float)
        +removeMoney(player: Player, amount: float)
        +calculate_rake() : float
    }

    Hand --> Pot : contains
```


the logger, clearly the idea is to activate different types of level on different parts of the code to track bugs ... and above all recentralize it in a single config file (or at least centralize it in the same place)


for configuration, I'd like to use a more modern format like yaml or toml, but that would mean rewriting the whole configuration.py file, plus I'd like to have hot loading of configuration changes

Thanks for your constructive and motivating feedback.

**Comment by brunetton** [comment]
Hi @jejellyroll-fr, here I am again, struggling with the logging on this project. The `loggingFpdb.py` is 1600 lines long (:astonished:) and I don't understand much about it. Are you sure you haven't reinvented the wheel ?

I'm tracking down the problem in #101 but to start, the bare minimum is being able to see debug logs in console. But I didn't get howto do this here :( Why is this so different from standard way of logging in Python ? 

Could you write down a little guide somewhere (in the wiki for example) to quickly explain how to activate DEBUG logging for a particular logger (`hud` or `hud_main` in my case, I'm not sure but I'll find out) for the console ? (not the GUI as I can't grep, copy/paste, redirect to file etc from GUI)

Thanks !

**Comment by jejellyroll-fr** [comment]
hello @brunetton 

Thank you for your feedback.
It's a pragmatic choice. As I mentioned, I work on several VMs (Linux X11/Wayland, Windows, macOS) without a centralised stack for collecting logs. I need to be able to quickly target the logs of a specific module in the console. 
If you have any simpler suggestions for bringing the API closer to the standard logging library while retaining this fine filtering by module in the console, please feel free to submit a PR.

**Reply by brunetton** [reply]
> It's a pragmatic choice. As I mentioned, I work on several VMs (Linux X11/Wayland, Windows, macOS) without a centralised stack for collecting logs. I need to be able to quickly target the logs of a specific module in the console.

Yes, quickly activate a module's logs is a classic need; but I don't really see why it would change from one OS to another. As soon as there is a console, the log system should behave identically (or I'm missing something). One classic method is the method used by the project: using a separate ini file to configure the levels of each logger: https://michaelheap.com/using-ini-config-with-python-logger

There are also quite a bunch of libs that add granularity to control; for example this one made by my friend: https://daiquiri.readthedocs.io

But anyway, ***I really don't want to be intrusive or worse: wasting your time***. I'm conscient that the technical debt is huge and the time is not payed !

Maybe you missed this part of my comment:

> Could you write down a little guide somewhere (in the wiki for example) to quickly explain how to activate DEBUG logging for a particular logger (hud or hud_main in my case, I'm not sure but I'll find out) for the console ? (not the GUI as I can't grep, copy/paste, redirect to file etc from GUI)

If you're okay with the logging you've done (I mean it won't change a lot in the future), could you quickly explain howto activate DEBUG logging for a particular module ? 

Thanks

**Reply by brunetton** [reply]
After some tries I ended up with: 

```python
from loggingFpdb import get_logger_registry
import logging

get_logger_registry().set_logger_level("hud_main", logging.DEBUG)
```

But this enables the DEBUG level for all modules :(

**Reply by jejellyroll-fr** [reply]
weird!! i need to recheck this

**Reply by brunetton** [reply]
> weird!! i need to recheck this

And even weird: when I remove this, the DEBUG still remains for all the loggers ! I didn't found a loggin.conf file or similar (maybe I missed it, it was late yesterday). I don't think it's written to the database (?) 

Once again, I may be wrong but I really have the feeling that you maybe reinvented the wheel here, and added quite a big technical dept by not using something community maintained (tested, working, stable). But I didn't understood the idea behind this code (the AI I gave used to help me didn't understood the idea too)
